modules/core/04-channel/keeper/upgrade.go

```diff
@@ -265,3 +264,61 @@ func (k Keeper) constructProposedUpgrade(ctx sdk.Context, portID, channelID stri
 		LatestSequenceSend: seq - 1,
 	}, nil
 }
+
+
+// abortHandshake will restore the channel state and flush status to their pre-upgrade state so that upgrade is aborted.
+// any unnecessary state is deleted. An error receipt is written, and the OnChanUpgradeRestore callback is called.
+func (k Keeper) abortHandshake(ctx sdk.Context, portID, channelID string, upgradeError *types.UpgradeError) error {
+	upgrade, found := k.GetUpgrade(ctx, portID, channelID)
+	if !found {
+		return errorsmod.Wrapf(types.ErrUpgradeNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if err := k.restoreChannel(ctx, portID, channelID); err != nil {
+		return err
+	}
+
+	// TODO: could we want to abort but have upgrade error be non-nil?
+	if upgradeError != nil {
+		if err := k.writeErrorReceipt(ctx, portID, channelID, upgrade, upgradeError); err != nil {
+			return err
+		}
+	}
+
+	// TODO: callback execution
+	// cbs.OnChanUpgradeRestore()
```

Comment:
this likely won't be here, this will be at the message server layer, I can remove this TODO if it's confusing